symengine/printer.cpp

```diff
@@ -57,6 +57,21 @@ void StrPrinter::bvisit(const NaN &x)
     str_ = s.str();
 }
 
+void StrPrinter::bvisit(const Factorial &x)
+{
+    std::ostringstream s;
+    s << "(" << apply(x.get_arg()) << ")!";
```

Comment:
`x!` should not be printed as `(x)!`

Reply:
This was actually done for [these](https://github.com/symengine/symengine/pull/1337/files/7ba71508d9f310d87f9840a87270f133a7d915b5#diff-971819d8abcc9340e783c26a4a53a5cfR569) cases.

Reply:
I mean you should check the precedence of the arg before printing the parenthesis. Take a look at how parenthesis are added. For example, `x**n` vs `(x+y)**n` where there is no parenthesis around `x` in the first case